packages/autocertifier-client/src/AutoCertifierClient.ts

```diff
@@ -0,0 +1,172 @@
+}
+
+const logger = new Logger(module)
+
+export const AUTOCERTIFIER_SERVICE_ID = 'system/auto-certificer'
+const ONE_DAY = 1000 * 60 * 60 * 24
+const MAX_INT_32 = 2147483647
+
+export class AutoCertifierClient extends EventEmitter<AutoCertifierClientEvents> implements IAutoCertifierService {
+
+    private updateTimeout?: NodeJS.Timeout
+    private readonly restClient: RestClient
+    private readonly subdomainPath: string
+    private readonly streamrWebSocketPort: number
+    private readonly ongoingSessions: Set<string> = new Set()
+
+    constructor(
+        subdomainPath: string,
+        streamrWebSocketPort: number,
+        restApiUrl: string,
+        registerRpcMethod: (
+            serviceId: string,
+            rpcMethodName: string,
+            method: (request: SessionIdRequest, context: ServerCallContext) => Promise<SessionIdResponse>
+        ) => void
+    ) {
+        super()
+
+        this.restClient = new RestClient(restApiUrl)
+        this.subdomainPath = filePathToNodeFormat(subdomainPath)
+        this.streamrWebSocketPort = streamrWebSocketPort
+        registerRpcMethod(AUTOCERTIFIER_SERVICE_ID, 'getSessionId', this.getSessionId.bind(this))
+    }
+
+    public async start(): Promise<void> {
+        if (!fs.existsSync(this.subdomainPath)) {
+            await this.createCertificate()
+        } else {
+            await this.checkSubdomainValidity()
+        }
+    }
+
+    private async checkSubdomainValidity(): Promise<void> {
+        const sub = this.loadSubdomainFromDisk()
+
+        if (Date.now() >= sub.expirationTimestamp - ONE_DAY) {
+            await this.updateCertificate()
+        } else {
+            await this.updateSubdomainIpAndPort()
+            this.scheduleCertificateUpdate(sub.expirationTimestamp)
+            this.emit('updatedSubdomain', sub.subdomain)
+        }
+    }
+
+    private loadSubdomainFromDisk(): { subdomain: CertifiedSubdomain, expirationTimestamp: number } {
+        const subdomain = JSON.parse(fs.readFileSync(this.subdomainPath, 'utf8')) as CertifiedSubdomain
+        const certObj = forge.pki.certificateFromPem(subdomain.certificate.cert)
+        const expirationTimestamp = certObj.validity.notAfter.getTime()
+        return { subdomain, expirationTimestamp }
+    }
+
+    public stop(): void {
+        if (this.updateTimeout) {
+            clearTimeout(this.updateTimeout)
+            this.updateTimeout = undefined
+        }
+    }
+
+    private scheduleCertificateUpdate(expirationTimestamp: number): void {
+        if (this.updateTimeout) {
+            clearTimeout(this.updateTimeout)
+            this.updateTimeout = undefined
+        }
+        // update certificate 1 day before it expires
+        let updateIn = expirationTimestamp - Date.now()
+        if (updateIn > ONE_DAY) {
+            updateIn = updateIn - ONE_DAY
+        }
+
+        if (updateIn > MAX_INT_32) {
+            updateIn = MAX_INT_32
+        }
```

Comment:
Can this happen in practice?

Reply:
Apparently so, which is why it was implemented

Reply:
Does this mean that quite typically we update the certificate already after 24.8 days (`MAX_INT_32` / `ONE_DAY`) ?

E.g. the Let's encrypt certificates  are usually valid for 90 days. Are the certificates created by the `autocertifier` also valid ~90 days?